### PR TITLE
[portmgr] Fixed the orchagent crash due to late arrival of notif

### PR DESCRIPTION
**Pull request by vivekrnv**
**What I did**

1) Bulk write to APP_DB i.e. alias, lanes, speed must be read through one notification by orchagent during create_port
2) Handled a race condition in portmgrd which tries to immediately apply a mtu/admin_status SET notif after a DEL causing it to crash

**Why I did it**

After the recent incremental config update changes, portmgrd will be writing to the APPL_DB PORT_TABLE (Except during startup where portyncd writes to the APP_DB). 

This can lead to a problem during breakout/breakin and might cause orchagent to crash because portmgrd currently doesn't write to the APPL_DB at once but in steps. 

In some cases (high probability, not always), orchagent decides to act on the notifcations immediately and since speed is not set, it tries to create_port with speed 0 and the orchagent crashes 

```
2022-08-25.10:45:56.708466|PORT_TABLE:Ethernet232|DEL
2022-08-25.10:45:57.669534|PORT_TABLE:Ethernet232|SET|alias:etp30a|index:30
2022-08-25.10:45:57.670749|PORT_TABLE:Ethernet232|SET|lanes:232,233
```

```
Aug 25 13:45:57.838085 r-leopard-41 ERR syncd#SDK: [SAI_PORT.ERR] mlnx_sai_port.c[2809]- mlnx_port_speeds_merge: Port [Speed[0x00000000] - Interface_Type[0x0000ffff]] configuration is invalid.
Aug 25 13:45:57.838085 r-leopard-41 ERR syncd#SDK: [SAI_PORT.ERR] mlnx_sai_port.c[3086]- mlnx_port_update_speed: Failed to merge speeds, interface types and auto_neg.
Aug 25 13:45:57.838085 r-leopard-41 ERR syncd#SDK: [SAI_PORT.ERR] mlnx_sai_port.c[9666]- mlnx_create_port: Failed to update speed.
Aug 25 13:45:57.838674 r-leopard-41 ERR swss#orchagent: :- create: create status: SAI_STATUS_NOT_SUPPORTED
Aug 25 13:45:57.838674 r-leopard-41 ERR swss#orchagent: :- addPort: Failed to create port with the speed 0, rv:-2
Aug 25 13:45:57.838693 r-leopard-41 ERR swss#orchagent: :- handleSaiCreateStatus: Encountered failure in create operation, exiting orchagent, SAI API: SAI_API_PORT, status: SAI_STATUS_NOT_SUPPORTED
```

```
2022-08-25.10:45:57.671548|c|SAI_OBJECT_TYPE_PORT:oid:0x1000000000adb|SAI_PORT_ATTR_SPEED=0|SAI_PORT_ATTR_HW_LANE_LIST=2:232,233
2022-08-25.10:45:57.838523|E|SAI_STATUS_NOT_SUPPORTED
```

Also handled a race condition in portmgrd happens because it was trying to set mtu/admin_status immediately after the a DEL was sent. portmgrd raced to set the fields on a netdev which was by then deleted and thus portmgrd crashed

**How I verified it**

1) Existing Unit Tests
2) Tested on the switch:

Before the fix:

root@r-panther-23:/home/admin# config interface breakout Ethernet72 2x50G[25G,10G] -y -f

```
2022-08-26.03:57:26.451022|INTF_TABLE:Ethernet72|DEL
2022-08-26.03:57:27.208461|PORT_TABLE:Ethernet72|SET|alias:etp19a|index:19
2022-08-26.03:57:27.211031|PORT_TABLE:Ethernet72|SET|lanes:72,73|speed:50000
2022-08-26.03:57:27.852567|PORT_TABLE:Ethernet72|SET|mtu:9100|admin_status:down
2022-08-26.03:57:28.245470|PORT_TABLE:Ethernet72|SET|mtu:9100
2022-08-26.03:57:28.250334|PORT_TABLE:Ethernet72|SET|admin_status:down
```

After the fix:
```
2022-08-26.04:06:22.060385|PORT_TABLE:Ethernet72|SET|alias:etp19a|index:19|lanes:72,73|speed:50000|mtu:9100
2022-08-26.04:06:22.865474|PORT_TABLE:Ethernet72|SET|admin_status:down
```

**Details if related**

### CODE REVIEW

**Review by dgsudharsan — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@vivekrnv could you please explain why we need sleep of 1 sec in test between each iteration? or why only 1 sec in enought?
@prsunny please help or assign someone to review.

**Review by dgsudharsan — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@Junchao-Mellanox  FYI

File: cfgmgr/portmgr.cpp [hunk 1]
```diff
@@ -124,10 +152,9 @@ void PortMgr::doTask(Consumer &consumer)
                 }
             }
 
-            for (auto &entry : field_values)
+            if (field_values.size())
             {
-                writeConfigToAppDb(alias, fvField(entry), fvValue(entry));
-                SWSS_LOG_NOTICE("Configure %s %s to %s", alias.c_str(), fvField(entry).c_str(), fvValue(entry).c_str());
+                writeConfigToAppDb(alias, field_values);
```
Comment (Junchao-Mellanox) [inline cfgmgr/portmgr.cpp:157]: Where to put the notice log?

File: cfgmgr/portmgr.cpp [hunk 2]
```diff
@@ -23,27 +23,55 @@ PortMgr::PortMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, c
...
-    return writeConfigToAppDb(alias, "admin_status", (up ? "up" : "down"));
+    cmd_str = cmd.str();
+    int ret = swss::exec(cmd_str, res);
+    if (!ret)
+    {
+        return writeConfigToAppDb(alias, "admin_status", (up ? "up" : "down"));
+    }
+    else if (!isPortStateOk(alias))
+    {
+        // Can happen when a DEL notification is sent by portmgrd followed by a pending admin_status SET notification
```
Comment (Junchao-Mellanox) [inline cfgmgr/portmgr.cpp:66]: Not sure I understand the comment, could you please elaborate?

File: cfgmgr/portmgr.cpp [hunk 3]
```diff
@@ -176,3 +204,9 @@ bool PortMgr::writeConfigToAppDb(const std::string &alias, const std::string &fi
 
     return true;
 }
+
+bool PortMgr::writeConfigToAppDb(const std::string &alias, std::vector<FieldValueTuple> &field_values)
+{
+    m_appPortTable.set(alias, field_values);
+    return true;
```
Comment (Junchao-Mellanox) [inline cfgmgr/portmgr.cpp:211]: do we need the return value?

File: tests/port_dpb.py [hunk 4]
```diff
@@ -242,8 +242,10 @@ def breakin(self, dvs, port_names):
 
         for cp in child_ports:
             assert(cp.exists_in_config_db() == False)
+        time.sleep(1)
```
Comment (Junchao-Mellanox) [inline tests/port_dpb.py:245]: Why do we need the sleep here?

File: cfgmgr/portmgr.cpp (hunk 1 above)
Comment (vivekrnv) [reply cfgmgr/portmgr.cpp:157]: @prsunny recommended to remove it

File: cfgmgr/portmgr.cpp (hunk 2 above)
Comment (vivekrnv) [reply cfgmgr/portmgr.cpp:66]: portmgrd defers adds the admin_status/mtu notifs back to the consumer queue when the netdev is not created. So, let's assume, there is a DEL notif which arrived after the first SET, We'll process DEL first and then followed by the deferred admin_status/mtu notif. 

If this happens there might be a possibility where portmgrd tries to set mtu/admin_status on a deleted netdev iface and in process throws a runtime error.

File: cfgmgr/portmgr.cpp (hunk 3 above)
Comment (vivekrnv) [reply cfgmgr/portmgr.cpp:211]: The other func (writeConfigToAppDb) has one. To keep it consistent, i added it here

File: tests/port_dpb.py (hunk 4 above)
Comment (vivekrnv) [reply tests/port_dpb.py:245]: I've saw this failing sometime and the reason i thought was this https://github.com/sonic-net/sonic-swss/pull/2431#issuecomment-1230499090. The idea is to stabilize

File: cfgmgr/portmgr.cpp (hunk 2 above)
Comment (Junchao-Mellanox) [reply cfgmgr/portmgr.cpp:66]: I am not sure that if it is possible. I suppose DEL will override everything.  Could you please check this https://github.com/sonic-net/sonic-swss/blob/ac8a83f0cd36c11c1a480a682e4eabbd99bf962c/orchagent/orch.cpp#L100 ?

File: tests/port_dpb.py (hunk 4 above)
Comment (Junchao-Mellanox) [reply tests/port_dpb.py:245]: In https://github.com/sonic-net/sonic-swss/pull/2431#issuecomment-1230499090, you mentioned that it did not help...

Comment (Junchao-Mellanox) [reply tests/port_dpb.py:245]: Is it a new test case? Why was there no sleep before?

File: cfgmgr/portmgr.cpp (hunk 2 above)
Comment (vivekrnv) [reply cfgmgr/portmgr.cpp:66]: Oh, i'm sorry. I was interpreting it wrong.  

A DEL came first and before the STATE_DB got updated, a SET came with all the fields.
```
2022-09-06.01:30:50.803955|PORT_TABLE:Ethernet64|DEL
2022-09-06.01:30:50.841680|PORT_TABLE:Ethernet64|SET|alias:fortyGigE0/64|index:16|lanes:69,70,71,72|speed:100000
2022-09-06.01:30:51.854363|PORT_TABLE:Ethernet64|SET|mtu:9100
2022-09-06.01:30:51.860562|PORT_TABLE:Ethernet64|SET|admin_status:down
```

But the scenario certainly happened: This was the log in one of the tests:
```
2022-09-01T09:22:37.2380593Z ______________________ TestPortDPB.test_port_breakout_one ______________________
2022-09-01T09:22:37.2381024Z 
2022-09-01T09:22:37.2381526Z self = <test_port_dpb.TestPortDPB object at 0x7f90ce7d0850>
2022-09-01T09:22:37.2382156Z dvs = <conftest.DockerVirtualSwitch object at 0x7f90ce82fe80>
2022-09-01T09:22:37.2382620Z 
2022-09-01T09:22:37.2383089Z     def test_port_breakout_one(self, dvs):
2022-09-01T09:22:37.2383573Z         dpb = DPB()
2022-09-01T09:22:37.2384079Z >       dpb.breakout(dvs, "Ethernet0", maxBreakOut)
2022-09-01T09:22:37.2384424Z 
2022-09-01T09:22:37.2384862Z test_port_dpb.py:45: 
2022-09-01T09:22:37.2385414Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-09-01T09:22:37.2385981Z port_dpb.py:304: in breakout
2022-09-01T09:22:37.2386534Z     self.create_child_ports(dvs, p, num_child_ports)
2022-09-01T09:22:37.2387143Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-09-01T09:22:37.2387503Z 
2022-09-01T09:22:37.2387967Z self = <port_dpb.DPB object at 0x7f90cece7af0>
2022-09-01T09:22:37.2388571Z dvs = <conftest.DockerVirtualSwitch object at 0x7f90ce82fe80>
2022-09-01T09:22:37.2389216Z p = <port_dpb.Port object at 0x7f90cece7730>, num_child_ports = 4
2022-09-01T09:22:37.2389605Z 
2022-09-01T09:22:37.2390103Z     def create_child_ports(self, dvs, p, num_child_ports):
2022-09-01T09:22:37.2390697Z         # Create child ports and write to config DB
2022-09-01T09:22:37.2391263Z         child_ports = p.port_split(num_child_ports)
2022-09-01T09:22:37.2391792Z         child_port_names = []
2022-09-01T09:22:37.2392263Z         for cp in child_ports:
2022-09-01T09:22:37.2392754Z             cp.write_to_config_db()
2022-09-01T09:22:37.2393279Z             child_port_names.append(cp.get_name())
2022-09-01T09:22:37.2393897Z         #print "Added child ports:%s to config DB"%child_port_names
2022-09-01T09:22:37.2394413Z     
2022-09-01T09:22:37.2394843Z         for cp in child_ports:
2022-09-01T09:22:37.2395361Z             assert(cp.exists_in_config_db() == True)
2022-09-01T09:22:37.2395889Z             cp.verify_config_db()
2022-09-01T09:22:37.2396501Z         #print "Config DB verification passed"
2022-09-01T09:22:37.2397019Z         for cp in child_ports:
2022-09-01T09:22:37.2397536Z >           assert(cp.exists_in_app_db() == True)
2022-09-01T09:22:37.2398045Z E           AssertionError
2022-09-01T09:22:37.2398333Z 
2022-09-01T09:22:37.2398778Z port_dpb.py:279: AssertionError
```

This happened for first child port, portmgrd crashed and thus nothing was written to APP_DB for other child ports. i've checked in my local vs and saw the portmgrd runtime error log as well

File: tests/port_dpb.py (hunk 4 above)
Comment (vivekrnv) [reply tests/port_dpb.py:245]: Will remove it, i guess this is not necessary

Comment (vivekrnv) [reply tests/port_dpb.py]: Removed

**Review by Junchao-Mellanox — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@prsunny  as this is a degradation inteorudced with prev PRs i would appreciate if you can prioritize the review so we will have it on 202205

**Review by dgsudharsan — APPROVED** [review]

**Review by prsunny — COMMENTED** [review]
lgtm, @prgeor , can you please review?

File: orchagent/portsorch.cpp [hunk 5]
```diff
@@ -2615,6 +2615,15 @@ bool PortsOrch::addPort(const set<int> &lane_set, uint32_t speed, int an, string
 {
     SWSS_LOG_ENTER();
 
+    if (!speed || lane_set.empty())
```
Comment (prsunny) [inline orchagent/portsorch.cpp:2618]: Does this impact VS swss initialization?

**Review by prsunny — APPROVED** [review]

File: orchagent/portsorch.cpp (hunk 5 above)
Comment (vivekrnv) [reply orchagent/portsorch.cpp:2618]: That is why we are deferring until both of them are available. No, it doesn't. cfg used in VS has speed and lane_set defined, so shouldn't be a problem.

**Pull request merged**